crates/cdk/src/wallet/proofs.rs

```diff
@@ -166,112 +167,393 @@ impl Wallet {
-                next_bigger.clone()
+        if opts.method == ProofSelectionMethod::Least {
+            let selected_amount = Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
+            let keyset_fees = if opts.include_fees {
+                self.get_mint_keysets()
+                    .await?
+                    .into_iter()
+                    .map(|k| (k.id, k.input_fee_ppk))
+                    .collect()
             } else {
-                break;
+                HashMap::new()
             };
+            if let Some(proofs) = select_least_proofs_over_amount(
+                &active_proofs,
+                amount.checked_sub(selected_amount).unwrap_or(Amount::ZERO),
+                keyset_fees,
+            ) {
+                selected_proofs.extend(proofs);
+                return Ok(selected_proofs);
+            }
+        }
 
-            let proof_amount = selected_proof.amount;
-
-            selected_proofs.push(selected_proof);
+        sort_proofs(&mut active_proofs, opts.method, amount);
 
-            let fees = match include_fees {
-                true => self.get_proofs_fee(&selected_proofs).await?,
-                false => Amount::ZERO,
+        for active_proof in active_proofs {
+            selected_proofs.push(active_proof);
+            let selected_total = Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
+            let fees = if opts.include_fees {
+                self.get_proofs_fee(&selected_proofs).await?
+            } else {
+                Amount::ZERO
             };
 
-            if proof_amount >= remaining_amount + fees {
-                remaining_amount = Amount::ZERO;
-                break;
+            if selected_total >= amount + fees {
+                return Ok(selected_proofs);
             }
+        }
+
+        Err(Error::InsufficientFunds)
+    }
+}
 
-            remaining_amount = amount.checked_add(fees).ok_or(Error::AmountOverflow)?
-                - Amount::try_sum(selected_proofs.iter().map(|p| p.amount))?;
-            (proofs_larger, proofs_smaller) = proofs_smaller
-                .into_iter()
-                .skip(1)
-                .partition(|p| p.amount > remaining_amount);
+fn sort_proofs(proofs: &mut Proofs, method: ProofSelectionMethod, amount: Amount) {
+    match method {
+        // Least fallback to largest
+        ProofSelectionMethod::Largest | ProofSelectionMethod::Least => {
+            proofs.sort_by(|a: &Proof, b: &Proof| b.cmp(a))
         }
+        ProofSelectionMethod::Closest => proofs.sort_by(|a: &Proof, b: &Proof| {
+            let a_diff = if a.amount > amount {
+                a.amount - amount
+            } else {
+                amount - a.amount
+            };
+            let b_diff = if b.amount > amount {
+                b.amount - amount
+            } else {
+                amount - b.amount
+            };
+            a_diff.cmp(&b_diff)
+        }),
```

Comment:
Maybe this can be simplified to
```rust
ProofSelectionMethod::Closest => proofs.sort_by_key(|p| match p.amount > amount {
    true => p.amount - amount,
    false => amount - p.amount,
}),
```

Reply:
I'm not a huge fan of matching booleans since I think it reads less clearly, but I like using `sort_by_key`. Thanks!